Overhaul CONTRIBUTING.md (#7471)
near/nearcore: CONTRIBUTING.md

Thank you for your interest in contributing to the NEAR reference client!  We
welcome contributions from everyone.  Below are various bits of information to
help you get started.  If you require additional help, please reach out to us on
our [zulip channel](https://near.zulipchat.com/).

# Getting started

- For non-technical contributions, such as e.g. content or events, see [this
document](https://docs.nearprotocol.com/docs/contribution/contribution-overview).

- For an overview of the NEAR core architecture, see [this
playlist](https://www.youtube.com/playlist?list=PL9tzQn_TEuFV4qlts0tVgndnytFs4QSYo).

- If you have never contributed to nearcore before and are looking for
relatively simple issues to get started with, please checkout issues labeled
with `C-good-first-issue`
[here](https://github.com/near/nearcore/labels/C-good-first-issue).  If you see

> one that looks interesting and is not claimed, please comment on the issue that

Reviewer: I would consider removing ‘and is not claimed’.  We may have issues which are claimed but the author haven’t touched it for months and presumably abandoned it.

you would like to start working on it and someone from the team should assign it
to you.

- If you have an idea for an enhancement to the protocol, please make the
proposal by following the [NEAR Enhancement
Proposal](https://github.com/near/NEPs/blob/master/neps/nep-0001.md).

- Otherwise, if you have an idea for an enhancement that does not require a
change in the proposal, please start by creaeting a new issue in the repository.
Someone should respond within 48 hours.  If the proposal is accepted, then you
can follow the process below to begin working on it.

- Finally, if you have noticed an obvious typo or bug that can be fixed with a
small code change, please feel free to submit a PR directly addressing the issue
without opening an issue.

If you have an issue assigned to you, then we kindly ask that you provide an
update at least once a week so we know if you are still actively working on the
issue.

> # Setting up editors and IDEs

Reviewer 2: Extreme HTML pedantry, but `#` generates an `h1`, and there should be only one `h1` in an article. You want to use `##` for section titeles, and resreve `#` for overall titele of the document.

Majority of NEAR developers use CLion with Rust plugin as their primary IDE.

We also had success with VSCode with rust-analyzer, see the steps for
installation
[here](https://commonwealth.im/near/proposal/discussion/338-remote-development-with-vscode-and-rustanalyzer).

> Some of us use VIM with [rust.vim](https://github.com/rust-lang/rust.vim) and

Reviewer: I feel attacked by Emacs not being mentioned. ;)

Reviewer 2: Yeah... I wouldn't be against dropping this section altogether. We basically use "standard rust tooling", it's not helpful to belabor that. It also can be misleading. For vim, for example, I think the current best source of docs is https://sharksforarms.dev/posts/neovim-rust/ (and, for Emach, https://robert.kra.hn/posts/rust-emacs-setup/, although even that doesn't cover eglot)

Author: Agreed, removed the whole section.

[rusty-tags](https://github.com/dan-t/rusty-tags). It has fewer features than
CLion or VSCode, but overall provides a usable setting.

Refer to [this
document](https://docs.nearprotocol.com/docs/contribution/nearcore) for details
on setting up your environment.

# Build Process

Nearcore is a reasonably standard Rust project, so `cargo test` most likely will
just work.  There are couple of specifics though:

* `nearcore` assumes a UNIX-like environment.  Linux is actively supported; Mac
is generally supported (the newer M1 chipset may not be well supported); and
Windows is not supported.
* `nearcore` build process includes generating bindings to RocksDB via
`bindgen`, which requires `libclang`.  See [`bindgen`
documentation](https://rust-lang.github.io/rust-bindgen/requirements.html#clang)
for installation instructions.

> * At the moment, `nearcore` only supports compiling for x86_64 architecture (for
> technical reasons, we have to care about CPU architecture and supporting one is
> easier).  To compile to x86_64 on Apple silicon, use `rustup set default-host
> x86_64-apple-darwin` command.  That is, cross-compilation sadly doesn't work at
> the time of writing.

Reviewer 2: ```suggestion
```

We've fixed this!

* You can optionally use the system installations of `librocksdb`, `libsnappy`
and `lz4` in order to speed up the compilation and reduce the build memory

> requirements by setting the `ROCKSDB_LIB_DIR`, `SNAPPY_LIB_DIR` and

Reviewer: Should we link to https://wiki.near.org/contribute/contribute-nearcore#use-system-provided-rocksdb

`LZ4_LIB_DIR` environment variables.  These environment variables should point at
the directory containing the dynamic shared objects (`.so`s or `.dylib`s.)

If your setup does not involve `rustup`, the required version of the rust
toolchain can be found in the `rust-toolchain` file.

# Pull Requests

All the contributions to `nearcore` happen via Pull Requests.  To create a Pull
Request, fork `nearcore`, create a new branch, do the work there, and then send
a PR against the `master` branch.

When working on a PR, please keep the following in mind:

1. The changes should be thoroughly tested.  Please refer to [this
document](https://github.com/nearprotocol/nearcore/wiki/Writing-tests-for-nearcore)
for our testing guidelines and overview of the testing infrastructure.
2. If you get an issue assigned to you, please post updates at least once a
week.  You are encouraged to submit work-in-progress draft PRs to get early

> feedback.

Reviewer 2: I suggest cutting this. We are not assigning people to issues/PRs consistenly, and we certainly do let PRs to linger for more than a week. 

To have this rule written down, we need someone to actually enforce it.

Author: Agreed, now removed.

3. A PR can contain any number of commits and when merged, the commits will be
squashed into a single commit.
4. The PR name should follow the template: `<type>: <name>`.  

> Where `type` is `fix` for fixesö `feat` for new features; `refactor` for changes

Reviewer: ```suggestion
Where `type` is `fix` for fixes; `feat` for new features; `refactor` for changes
```

that primarily reorganize code; `doc` for changes that primarily change
documentation or comments; `test` for changes that primarily introduce new
tests; and `chore` for updating grunt tasks. Please note that `type` is case

> sensitive.

Reviewer 2: Let's maybe itemize section titles for bettre readability: 

* `fix`
* `test`
* `feat`
* `chore`

5. The PR description should follow the following template:

```
<description>

Test plan
---------
<test plan>
```

> 6. If your PR introduces a new protocol feature, please document it in
> [CHANGELOG.md](CHANGELOG.md) under `unreleased`.

Reviewer: ```suggestion
6. If your PR introduces a user-observable change (e.g. a new protocol
feature, new configuration option, new Prometheus metric etc.) please
document it in [CHANGELOG.md](CHANGELOG.md) in `[unreleased]` section.
```

This process is managed by [cargo-workspaces](https://github.com/pksunkara/cargo-workspaces), with a [bit of magic](https://github.com/pksunkara/cargo-workspaces/compare/master...miraclx:grouping-and-exclusion#files_bucket) sprinkled on top.

## Issue Labels

Issue labels are of the following format `<type>-<content>` where `<type>` is a capital letter indicating the type of the label and `<content>` is a hyphened phrase indicating what is label is about.
For example, in the label `C-bug`, `C` means category and `bug` means that the label is about bugs.
Common types include `C`, which means category, `A`, which means area, `T`, which means team.

An issue can have multiple labels including which area it touches, which team should be responsible for the issue, and so on.

> Each issue should have at least one label attached to it after it is triaged and the label could be a general one, such as `C-enhancement` or `C-bug`.

Author: These two sections felt a bit out of place in this file so I dropped them.  I am happy to include them if someone feels strongly that they should be included here.

Reviewer 2: :+1: for trimming this out of CONTRIBUTING.md. CONTRIBUTING.md is a landing page, it should be short and to the point, and just link other documents. 

:-1: for removing this info -- this documents non-obvious institutional knowledge, it should be documented somewhere. It's probably best to find a place for it somewhere in the `docs/` folder. There isn't an obvious place to put it, so I suggest creating a `docs/README.md`. That'll get automatically displayed by github when opening a folder, and can serve both as an index and as dumping ground for stuff we are not quite sure where else to put.

Author: Moved to `docs/README.md`

## Pull Requests

All the contributions to `nearcore` happen via Pull Requests.  To create a Pull
Request, fork `nearcore`, create a new branch, do the work there, and then send
a PR against the `master` branch.

When working on a PR, please keep the following in mind:

1. The changes should be thoroughly tested.  Please refer to [this
document](https://github.com/nearprotocol/nearcore/wiki/Writing-tests-for-nearcore)
for our testing guidelines and overview of the testing infrastructure.
2. Feel free to submit draft PRs to get early feedback and to make sure you are
on the right track.
3. A PR can contain any number of commits and when merged, the commits will be
squashed into a single commit.
4. The PR name should follow the template: `<type>: <name>`.  
Where `type` is:
   - `fix` for bug fixes;
   - `feat` for new features;
   - `refactor` for changes that reorganize code without adding new content;
   - `doc` for changes that change documentation or comments;
   - `test` for changes that introduce new tests;
   - `chore` for grunt tasks like updating dependencies.
5. The PR description should follow the following template:

```
<description>

Test plan
---------
<test plan>

> ```

Reviewer 2: I'd want to bike shed on whether we need this rule or not. We are definitely not following it, so we should either start enforcing it, or to remove it. But this is out of scope for the PR.

Reviewer: I was also suspicious of this.

Author: Agreed, will remove.

6. If your PR introduces a user-observable change (e.g. a new protocol
feature, new configuration option, new Prometheus metric etc.) please
document it in [CHANGELOG.md](CHANGELOG.md) in `[unreleased]` section.

## After the PR is submitted

1. We have a CI process configured to run various tests on each PR.  All tests
need to pass before your PR can be merged.

> 2. When the CI is green and comments from the reviewer(s) have been addressed,
> they should approve the PR enabling you to merge it.

Reviewer: ```suggestion
2. When all the comments from the reviewer(s) have been addressed, they should
approve the PR enabling it to be merged.
```

3. You can use the `S-automerge` label to instruct Github to put the PR in the
merge queue with other approved PRs so you do not have to manually look for a

> window to merge it.

Reviewer 2: Let's rephrase this a bit maybe:

> To merge an approved PR, add `S-automerge` label. The label is added by PR's author if they have commit access, and by a reviewer otherwise. PR author can also apply automerge label immediaetly after filing a PR: this could remove one round-trip after PR is approved. 

Three bits of info: 

* it's on author to `automerge`
* except that if you don't have commit access
* document pre-emptive auto-merge trick

Reviewer: Commit access doesn’t tie into it though, does it? Anyone can apply the label, don’t they? It’s just that what usually happens is if we have external contributors they don’t know about it so we end up applying the label.  I’d just change the last point to:

```suggestion
3. You can use the `S-automerge` label to put the PR onto the merge queue with
other approved PRs.
```

Reviewer 2: >Commit access doesn’t tie into it though, does it?

I think it does: you need to have a triage permission to label issues and PR. Not exactly commit access, but I don't think an external contributor can label even their own PRs.

Author: Oh, I did not realise external folks cannot access labels.  I suppose it makes sense that they cannot do that.  I suppose if we had a system like [bors](https://github.com/bors-ng/bors-ng) that works via comments, then external folks should be able to merge their own PRs.  Not proposing that we switch but next time we are shopping, we should consider this issue.

I'll update the text.
